File: implementations/go/matrix/matrix.go
package matrix

import (
	"math/rand"
	"time"
)

// 初始化矩阵
func InitializeMatrix(n int) [][]float64 {
	rand.Seed(time.Now().UnixNano())
	matrix := make([][]float64, n)
	
	for i := 0; i < n; i++ {
		matrix[i] = make([]float64, n)
		for j := 0; j < n; j++ {
			matrix[i][j] = rand.Float64()
		}
	}
	
	return matrix
}

// 标准矩阵乘法
func MatrixMultiply(a, b [][]float64) [][]float64 {
	n := len(a)
	c := make([][]float64, n)
	
	for i := 0; i < n; i++ {
		c[i] = make([]float64, n)
		for j := 0; j < n; j++ {
			for k := 0; k < n; k++ {
				c[i][j] += a[i][k] * b[k][j]
			}
		}
	}
	
	return c
}

// 分块矩阵乘法
func MatrixMultiplyBlocked(a, b [][]float64, blockSize int) [][]float64 {
	n := len(a)
	c := make([][]float64, n)
	
	for i := 0; i < n; i++ {
		c[i] = make([]float64, n)
	}
	
	// 分块矩阵乘法
	for i := 0; i < n; i += blockSize {
		for j := 0; j < n; j += blockSize {
			for k := 0; k < n; k += blockSize {
				// 处理当前块
				for ii := i; ii < min(i+blockSize, n); ii++ {
					for jj := j; jj < min(j+blockSize, n); jj++ {
						for kk := k; kk < min(k+blockSize, n); kk++ {
							c[ii][jj] += a[ii][kk] * b[kk][jj]
						}
					}
				}
			}
		}
	}
	
	return c
}

// 辅助函数：返回两个整数中的较小值
func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

File: implementations/rust/src/matrix.rs
use rand::Rng;

// 初始化矩阵
pub fn initialize_matrix(n: usize) -> Vec<Vec<f64>> {
    let mut rng = rand::thread_rng();
    let mut matrix = vec![vec![0.0; n]; n];
    
    for i in 0..n {
        for j in 0..n {
            matrix[i][j] = rng.gen::<f64>();
        }
    }
    
    matrix
}

// 标准矩阵乘法
pub fn matrix_multiply(a: &Vec<Vec<f64>>, b: &Vec<Vec<f64>>) -> Vec<Vec<f64>> {
    let n = a.len();
    let mut c = vec![vec![0.0; n]; n];
    
    for i in 0..n {
        for j in 0..n {
            for k in 0..n {
                c[i][j] += a[i][k] * b[k][j];
            }
        }
    }
    
    c
}

// 分块矩阵乘法
pub fn matrix_multiply_blocked(a: &Vec<Vec<f64>>, b: &Vec<Vec<f64>>, block_size: usize) -> Vec<Vec<f64>> {
    let n = a.len();
    let mut c = vec![vec![0.0; n]; n];
    
    // 分块矩阵乘法
    for i in (0..n).step_by(block_size) {
        for j in (0..n).step_by(block_size) {
            for k in (0..n).step_by(block_size) {
                // 处理当前块
                for ii in i..std::cmp::min(i + block_size, n) {
                    for jj in j..std::cmp::min(j + block_size, n) {
                        for kk in k..std::cmp::min(k + block_size, n) {
                            c[ii][jj] += a[ii][kk] * b[kk][jj];
                        }
                    }
                }
            }
        }
    }
    
    c
}
